processing/src/main/java/io/druid/query/groupby/GroupByQueryQueryToolChest.java

```diff
@@ -356,45 +355,13 @@ public String apply(DimensionSpec input)
       @Override
       public byte[] computeCacheKey(GroupByQuery query)
       {
-        final DimFilter dimFilter = query.getDimFilter();
-        final byte[] filterBytes = dimFilter == null ? new byte[]{} : dimFilter.getCacheKey();
-        final byte[] aggregatorBytes = QueryCacheHelper.computeAggregatorBytes(query.getAggregatorSpecs());
-        final byte[] granularityBytes = query.getGranularity().cacheKey();
-        final byte[][] dimensionsBytes = new byte[query.getDimensions().size()][];
-        int dimensionsBytesSize = 0;
-        int index = 0;
-        for (DimensionSpec dimension : query.getDimensions()) {
-          dimensionsBytes[index] = dimension.getCacheKey();
-          dimensionsBytesSize += dimensionsBytes[index].length;
-          ++index;
-        }
-        final byte[] havingBytes = query.getHavingSpec() == null ? new byte[]{} : query.getHavingSpec().getCacheKey();
-        final byte[] limitBytes = query.getLimitSpec().getCacheKey();
-
-        ByteBuffer buffer = ByteBuffer
-            .allocate(
-                2
-                + granularityBytes.length
-                + filterBytes.length
-                + aggregatorBytes.length
-                + dimensionsBytesSize
-                + havingBytes.length
-                + limitBytes.length
-            )
-            .put(GROUPBY_QUERY)
-            .put(CACHE_STRATEGY_VERSION)
-            .put(granularityBytes)
-            .put(filterBytes)
-            .put(aggregatorBytes);
-
-        for (byte[] dimensionsByte : dimensionsBytes) {
-          buffer.put(dimensionsByte);
-        }
-
-        return buffer
-            .put(havingBytes)
-            .put(limitBytes)
-            .array();
+        return new CacheKeyBuilder(GROUPBY_QUERY, CacheKeyBuilder.EMPTY_BYTES)
+            .appendByte(CACHE_STRATEGY_VERSION)
+            .appendCacheable(query.getGranularity())
+            .appendCacheable(query.getDimFilter())
+            .appendCacheableList(query.getAggregatorSpecs())
+            .appendCacheableList(query.getDimensions())
```

Comment:
The limit push down in #3873 is applied when merging the per-segment results, so taking the limit out of the cache key should be fine there as well